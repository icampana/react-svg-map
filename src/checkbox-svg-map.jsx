import React from 'react';
import PropTypes from 'prop-types';
import SVGMap from './svg-map';

class CheckboxSVGMap extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			selectedLocationIds: props.selectedLocationIds || []
		};

		this.svgMapRef = React.createRef();
		this.isLocationSelected = this.isLocationSelected.bind(this);
		this.handleLocationClick = this.handleLocationClick.bind(this);
		this.handleLocationKeyDown = this.handleLocationKeyDown.bind(this);
	}

	componentDidMount() {
		// To handle initial selection if provided
		if (this.props.onChange) {
			const initialSelectedLocations = this.state.selectedLocationIds
				.map(id => this.svgMapRef.current.querySelector(`#${id}`))
				.filter(Boolean);
			this.props.onChange(initialSelectedLocations);
		}
	}

	isLocationSelected(location) {
		return this.state.selectedLocationIds.includes(location.id);
	}

	toggleLocation(locationId) {
		this.setState(prevState => {
			const selectedLocationIds = [...prevState.selectedLocationIds];
			const locationIndex = selectedLocationIds.indexOf(locationId);

			if (locationIndex !== -1) {
				selectedLocationIds.splice(locationIndex, 1);
			} else {
				selectedLocationIds.push(locationId);
			}

			if (this.props.onChange) {
				const selectedLocations = selectedLocationIds
					.map(id => this.svgMapRef.current.querySelector(`#${id}`))
					.filter(Boolean);
				this.props.onChange(selectedLocations);
			}

			return { selectedLocationIds };
		});
	}

	handleLocationClick(event) {
		event.preventDefault();
		const locationId = event.target.id;
		this.toggleLocation(locationId);
	}

	handleLocationKeyDown(event) {
		if (event.keyCode === 32) { // Spacebar
			event.preventDefault();
			const locationId = event.target.id;
			this.toggleLocation(locationId);
		}
	}

	render() {
		return (
			<div ref={this.svgMapRef}>
				<SVGMap
					map={this.props.map}
					role="group"
					locationRole="checkbox"
					locationTabIndex="0"
					className={this.props.className}
					locationClassName={this.props.locationClassName}
					locationAriaLabel={this.props.locationAriaLabel}
					isLocationSelected={this.isLocationSelected}
					onLocationClick={this.handleLocationClick}
					onLocationKeyDown={this.handleLocationKeyDown}
					onLocationMouseOver={this.props.onLocationMouseOver}
					onLocationMouseOut={this.props.onLocationMouseOut}
					onLocationMouseMove={this.props.onLocationMouseMove}
					onLocationFocus={this.props.onLocationFocus}
					onLocationBlur={this.props.onLocationBlur}
					childrenBefore={this.props.childrenBefore}
					childrenAfter={this.props.childrenAfter}
				/>
			</div>
		);
	}
}

CheckboxSVGMap.propTypes = {
	selectedLocationIds: PropTypes.arrayOf(PropTypes.string),
	onChange: PropTypes.func,
	map: PropTypes.shape({
		viewBox: PropTypes.string.isRequired,
		locations: PropTypes.arrayOf(
			PropTypes.shape({
				path: PropTypes.string.isRequired,
				name: PropTypes.string,
				id: PropTypes.string
			})
		).isRequired,
		label: PropTypes.string
	}).isRequired,
	className: PropTypes.string,
	locationClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
	locationAriaLabel: PropTypes.func,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationMouseMove: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
	childrenBefore: PropTypes.node,
	childrenAfter: PropTypes.node,
};

export default CheckboxSVGMap;