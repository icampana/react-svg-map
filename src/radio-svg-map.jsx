import React from 'react';
import PropTypes from 'prop-types';
import SVGMap from './svg-map';

class RadioSVGMap extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			selectedLocationId: props.selectedLocationId || null
		};

		this.svgMapRef = React.createRef();
		this.locations = [];
		this.getLocationTabIndex = this.getLocationTabIndex.bind(this);
		this.isLocationSelected = this.isLocationSelected.bind(this);
		this.handleLocationClick = this.handleLocationClick.bind(this);
		this.handleLocationKeyDown = this.handleLocationKeyDown.bind(this);
	}

	componentDidMount() {
		this.locations = [...this.svgMapRef.current.querySelectorAll('path')];
		if (this.props.onChange && this.state.selectedLocationId) {
			const initialSelectedLocation = this.locations.find(loc => loc.id === this.state.selectedLocationId);
			if (initialSelectedLocation) {
				this.props.onChange(initialSelectedLocation);
			}
		}
	}

	getLocationTabIndex(location) {
		if (this.state.selectedLocationId) {
			return this.isLocationSelected(location) ? '0' : '-1';
		} else {
			// If no location is selected, make the first one focusable
			return this.props.map.locations[0].id === location.id ? '0' : '-1';
		}
	}

	isLocationSelected(location) {
		return this.state.selectedLocationId === location.id;
	}

	selectLocation(locationId) {
		const newSelectedLocation = this.locations.find(loc => loc.id === locationId);
		if (newSelectedLocation && newSelectedLocation.id !== this.state.selectedLocationId) {
			this.setState({ selectedLocationId: newSelectedLocation.id }, () => {
				newSelectedLocation.focus();
				if (this.props.onChange) {
					this.props.onChange(newSelectedLocation);
				}
			});
		}
	}

	handleLocationClick(event) {
		event.preventDefault();
		this.selectLocation(event.target.id);
	}

	handleLocationKeyDown(event) {
		const focusedLocationId = event.target.id;
		const focusedLocationIndex = this.locations.findIndex(loc => loc.id === focusedLocationId);

		if (event.keyCode === 32) { // Spacebar
			event.preventDefault();
			this.selectLocation(focusedLocationId);
		} else if (event.keyCode === 39 || event.keyCode === 40) { // Arrow down or right
			event.preventDefault();
			const nextLocationIndex = (focusedLocationIndex + 1) % this.locations.length;
			this.selectLocation(this.locations[nextLocationIndex].id);
		} else if (event.keyCode === 37 || event.keyCode === 38) { // Arrow up or left
			event.preventDefault();
			const prevLocationIndex = (focusedLocationIndex - 1 + this.locations.length) % this.locations.length;
			this.selectLocation(this.locations[prevLocationIndex].id);
		}
	}

	render() {
		return (
			<div ref={this.svgMapRef}>
				<SVGMap
					map={this.props.map}
					role="radiogroup"
					locationTabIndex={this.getLocationTabIndex}
					locationRole="radio"
					className={this.props.className}
					locationClassName={this.props.locationClassName}
					locationAriaLabel={this.props.locationAriaLabel}
					isLocationSelected={this.isLocationSelected}
					onLocationClick={this.handleLocationClick}
					onLocationKeyDown={this.handleLocationKeyDown}
					onLocationMouseOver={this.props.onLocationMouseOver}
					onLocationMouseOut={this.props.onLocationMouseOut}
					onLocationMouseMove={this.props.onLocationMouseMove}
					onLocationFocus={this.props.onLocationFocus}
					onLocationBlur={this.props.onLocationBlur}
				/>
			</div>
		);
	}
}

RadioSVGMap.propTypes = {
	selectedLocationId: PropTypes.string,
	onChange: PropTypes.func,
	map: PropTypes.shape({
		viewBox: PropTypes.string.isRequired,
		locations: PropTypes.arrayOf(
			PropTypes.shape({
				path: PropTypes.string.isRequired,
				name: PropTypes.string,
				id: PropTypes.string
			})
		).isRequired,
		label: PropTypes.string
	}).isRequired,
	className: PropTypes.string,
	locationClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
	locationAriaLabel: PropTypes.func,
	onLocationMouseOver: PropTypes.func,
	onLocationMouseOut: PropTypes.func,
	onLocationMouseMove: PropTypes.func,
	onLocationFocus: PropTypes.func,
	onLocationBlur: PropTypes.func,
	childrenBefore: PropTypes.node,
	childrenAfter: PropTypes.node,
};

export default RadioSVGMap;