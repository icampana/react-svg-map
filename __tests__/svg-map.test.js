import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import FakeMap from './fake-map';
import { SVGMap } from '../src/';

describe('SVGMap component', () => {
	describe('Properties', () => {
		test('displays map with default props', () => {
			const { asFragment } = render(<SVGMap map={FakeMap} />);
			expect(asFragment()).toMatchSnapshot();
		});

		test('displays map with custom props', () => {
			const eventHandler = jest.fn();
			const isLocationSelected = () => true;
			const { asFragment } = render(
				<SVGMap
					map={FakeMap}
					className="className"
					role="role"
					locationClassName="locationClassName"
					locationTabIndex="locationTabIndex"
					locationRole="locationRole"
					onLocationMouseOver={eventHandler}
					onLocationMouseOut={eventHandler}
					onLocationMouseMove={eventHandler}
					onLocationClick={eventHandler}
					onLocationKeyDown={eventHandler}
					onLocationFocus={eventHandler}
					onLocationBlur={eventHandler}
					isLocationSelected={isLocationSelected}
					childrenBefore={<text>childrenBefore</text>}
					childrenAfter={<text>childrenAfter</text>}
				/>
			);
			
			expect(asFragment()).toMatchSnapshot();
		});

		test('displays map with custom function location props', () => {
			const locationClassName = (location, index) => `locationClassName-${index}`;
			const locationTabIndex = (location, index) => `locationTabIndex-${index}`;
			const locationAriaLabel = (location, index) => `${location.name}-${index}`;
			const { asFragment } = render(
				<SVGMap
					map={FakeMap}
					locationClassName={locationClassName}
					locationTabIndex={locationTabIndex}
					locationAriaLabel={locationAriaLabel}
				/>
			);
			
			expect(asFragment()).toMatchSnapshot();
		});
	});

	describe('Events', () => {
		test('calls event handlers', () => {
			const onLocationMouseOver = jest.fn();
			const onLocationMouseOut = jest.fn();
			const onLocationMouseMove = jest.fn();
			const onLocationClick = jest.fn();
			const onLocationKeyDown = jest.fn();
			const onLocationFocus = jest.fn();
			const onLocationBlur = jest.fn();

			render(
				<SVGMap
					map={FakeMap}
					onLocationMouseOver={onLocationMouseOver}
					onLocationMouseOut={onLocationMouseOut}
					onLocationMouseMove={onLocationMouseMove}
					onLocationClick={onLocationClick}
					onLocationKeyDown={onLocationKeyDown}
					onLocationFocus={onLocationFocus}
					onLocationBlur={onLocationBlur}
				/>
			);

			const location = screen.getByRole('none', { name: 'name0' });

			fireEvent.mouseOver(location);
			expect(onLocationMouseOver).toHaveBeenCalledTimes(1);

			fireEvent.mouseOut(location);
			expect(onLocationMouseOut).toHaveBeenCalledTimes(1);

			fireEvent.mouseMove(location);
			expect(onLocationMouseMove).toHaveBeenCalledTimes(1);

			fireEvent.click(location);
			expect(onLocationClick).toHaveBeenCalledTimes(1);

			fireEvent.keyDown(location);
			expect(onLocationKeyDown).toHaveBeenCalledTimes(1);

			fireEvent.focus(location);
			expect(onLocationFocus).toHaveBeenCalledTimes(1);

			fireEvent.blur(location);
			expect(onLocationBlur).toHaveBeenCalledTimes(1);
		});
	});
});