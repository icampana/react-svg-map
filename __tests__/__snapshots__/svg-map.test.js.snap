// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGMap component Properties displays map with custom function location props 1`] = `
<DocumentFragment>
  <svg
    aria-label="label"
    class="svg-map"
    role="none"
    viewBox="viewBox"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      aria-label="name0-0"
      class="locationClassName-0"
      d="path0"
      id="id0"
      name="name0"
      role="none"
      tabindex="locationTabIndex-0"
    />
    <path
      aria-label="name1-1"
      class="locationClassName-1"
      d="path1"
      id="id1"
      name="name1"
      role="none"
      tabindex="locationTabIndex-1"
    />
    <path
      aria-label="name2-2"
      class="locationClassName-2"
      d="path2"
      id="id2"
      name="name2"
      role="none"
      tabindex="locationTabIndex-2"
    />
  </svg>
</DocumentFragment>
`;

exports[`SVGMap component Properties displays map with custom props 1`] = `
<DocumentFragment>
  <svg
    aria-label="label"
    class="className"
    role="role"
    viewBox="viewBox"
    xmlns="http://www.w3.org/2000/svg"
  >
    <text>
      childrenBefore
    </text>
    <path
      aria-checked="true"
      aria-label="name0"
      class="locationClassName"
      d="path0"
      id="id0"
      name="name0"
      role="locationRole"
      tabindex="locationTabIndex"
    />
    <path
      aria-checked="true"
      aria-label="name1"
      class="locationClassName"
      d="path1"
      id="id1"
      name="name1"
      role="locationRole"
      tabindex="locationTabIndex"
    />
    <path
      aria-checked="true"
      aria-label="name2"
      class="locationClassName"
      d="path2"
      id="id2"
      name="name2"
      role="locationRole"
      tabindex="locationTabIndex"
    />
    <text>
      childrenAfter
    </text>
  </svg>
</DocumentFragment>
`;

exports[`SVGMap component Properties displays map with default props 1`] = `
<DocumentFragment>
  <svg
    aria-label="label"
    class="svg-map"
    role="none"
    viewBox="viewBox"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      aria-label="name0"
      class="svg-map__location"
      d="path0"
      id="id0"
      name="name0"
      role="none"
      tabindex="0"
    />
    <path
      aria-label="name1"
      class="svg-map__location"
      d="path1"
      id="id1"
      name="name1"
      role="none"
      tabindex="0"
    />
    <path
      aria-label="name2"
      class="svg-map__location"
      d="path2"
      id="id2"
      name="name2"
      role="none"
      tabindex="0"
    />
  </svg>
</DocumentFragment>
`;
